Add shebang option to support more shells

danpf opened the pull request:
Previously the shell(shebang) was hardcoded to be '#!/bin/bash'.
This fixes that, because some of the queueing schedulers had their own
shebangs hardcoded as well, so you actually ended up having two shebang
lines.  In order to accomodate more shells, we can
add an option to specify shell via shebangs.  If you think this
implementation is fine, I can/will add this to the documentation as
well.

willirath commented on dask_jobqueue/core.py:
```diff
@@ -206,6 +207,8 @@ def __init__(self,
             env_extra = dask.config.get('jobqueue.%s.env-extra' % self.scheduler_name)
         if log_directory is None:
             log_directory = dask.config.get('jobqueue.%s.log-directory' % self.scheduler_name)
+        if shebang is None:
```
I'd reconsider adding a kwarg for this.  In my understanding, kwargs are for things you change frequently.  This looks like an issue that you have to sort out once for your deployment and then never touch it again.  Just adding it to the config should cover all cases.

guillaumeeb replied:
I'm not sure, and I have no strong feeling on this. Currently it looks like we have kind of a match between kwargs and config file. I'm not sure which way we should go.

danpf replied:
I did it this way only because that's how everything else in the yaml's are set.  I don't really mind either way.  maybe let me know @guillaumeeb ?

guillaumeeb replied:
Yeah, that is what I thought. If so, leave it as is, as the everything else is done. Maybe we'll need to check whether all the kwargs are justified or not in the future.

willirath commented on dask_jobqueue/jobqueue.yaml:
```diff
@@ -34,6 +35,7 @@ jobqueue:
     local-directory: null       # Location of fast local storage like /scratch or $TMPDIR
 
     # PBS resource manager options
+    shebang: "#!/bin/bash"
```
Shouldn't we go for `#!/usr/bin/env bash` ?

guillaumeeb replied:
Yep, looks like it, good catch!

willirath commented on dask_jobqueue/tests/test_slurm.py:
```diff
@@ -48,6 +48,7 @@ def test_job_script():
     with SLURMCluster(walltime='00:02:00', processes=4, cores=8, memory='28GB') as cluster:
 
         job_script = cluster.job_script()
+        assert job_script.startswith("#!/bin/bash")
```
This should also go into the tests for the other schedulers.

guillaumeeb replied:
I believe this should only go in https://github.com/dask/dask-jobqueue/blob/master/dask_jobqueue/tests/test_jobqueue_core.py as this is not specific to any scheduler.

willirath commented:
Looks good!

> Off Topic:
>
> Looking at this PR, adding all these identical lines to the configs (e.g., <https://github.com/dask/dask-jobqueue/pull/189/files#diff-06f4938c65f61f791782337bf4187054R15>) feels wrong.

guillaumeeb commented:
And thanks @willirath for the review here, this is really appreciated.

guillaumeeb commented on dask_jobqueue/tests/test_jobqueue_core.py:
```diff
@@ -42,6 +42,58 @@ def test_command_template():
         assert ' --preload mymodule' in cluster._command_template
 
 
+def test_shebang_settings():
```
You should use parametrized test as the one below for all this!

danpf replied:
really cool!!!!

guillaumeeb commented on dask_jobqueue/tests/test_slurm.py:
```diff
@@ -66,7 +66,7 @@ def test_job_script():
                       env_extra=['export LANG="en_US.utf8"',
                                  'export LANGUAGE="en_US.utf8"',
-                                 'export LC_ALL="en_US.utf8"']
+                                 'export LC_ALL="en_US.utf8"'],
```
A comma is remaining here.

guillaumeeb commented:
LGTM, @willirath do you see something else?

danpf commented:
sorry to bother you more, please review the updates to the documentation I added as well.
Also added some functionality to make sure there was always a shebang + test for that.

willirath commented:
Apart from the question about the added dependency on `mock`, this looks very good!

willirath commented on dask_jobqueue/tests/test_jobqueue_core.py:
```diff
@@ -42,6 +43,24 @@ def test_command_template():
+        assert job_script.startswith(default_shebang)
+
+    with pytest.raises(ValueError, match="batch script interpreter"):
+        with (dask.config.set(shebang=None)):
```
Sorry.  My example code was wrong.  `dask.config.set('jobqueue.%s.shebang' % Cluster.scheduler_name, None)` should work.

danpf replied:
thanks, my initial idea was way worse.

willirath commented:
> LGTM, @willirath do you see something else?

@guillaumeeb Good to go from py point of view!

guillaumeeb commented on docs/source/configuration-setup.rst:
```diff
@@ -8,7 +8,7 @@ You provide this information either as keyword arguments to the constructor:
 .. code-block:: python
 
-   cluster = PBSCluster(cores=36, memory='100GB', queue='regular', ...)
+   cluster = PBSCluster(cores=36, memory='100GB', shebang,"#!/usr/bin/env bash", queue='regular', ...)
```
Is there a comma instead of a equal sign here?
At the end, I would prefer to avoid exposing kwargs that are more rarely used everywhere in the docs. In this particular example I feel this is not needed.

danpf replied:
```suggestion
   cluster = PBSCluster(cores=36, memory='100GB', shebang="#!/usr/bin/env bash", queue='regular', ...)
```
great catch!

guillaumeeb commented on dask_jobqueue/core.py:
```diff
@@ -216,6 +221,10 @@ def __init__(self,
 
+        if shebang is None:
+            raise ValueError("You must specify a batch script interpreter to use"
+                             " like ``#!/usr/bin/env bash``")
```
Is this really a problem if there is no shebang? 
What concerns me here is that this exception might break existing jobqueue configurations, where no shebang is declared. From my point of view it shouldn't.

danpf replied:
We could maybe add a deprecation warning? and then a warning to say that if shebang is None everywhere, automatically set it to `#!/usr/bin/env bash`  to preserve the old configuration

You are right that this breaks existing configs in its current state.  But you cannot submit a batch script to most schedulers without a shebang:
```
sbatch: error: This does not look like a batch script.  The first
sbatch: error: line must start with #! followed by the path to an interpreter.
sbatch: error: For instance: #!/bin/sh
```

guillaumeeb replied:
Wait, I may be wrong about this. I need to try this PR to see if it breaks my config. Maybe it doesn't, default value should be taken from the module default jobqueue.yaml file.

guillaumeeb replied:
Sorry did not find time today, and this will not be possible until friday. You could verify it though, just create a jobqueue.yaml config file, and do not add a shebang line. 
I believe the only way to reach that code is to create a local config file, and set shebang to null in it. I'm thus in favor of removing this code, as it seems very unlikely a user run in this case without knowing it.

guillaumeeb commented on docs/source/configurations.rst:
```diff
@@ -103,6 +105,7 @@ ARM Stratus
         name: dask-worker
         cores: 36
         memory: 270GB
+        shebang: '#!/usr/bin/env bash'
```
You should remove the shebang from "real life" configurations, as keeping the default shebang is perfectly fine apparently for those. This include ARM, but also above Cheyenne and Nersc.

danpf replied:
Are these what the DOE are using as their `jobqueue.yaml` ? in that case they 'should' set the shebang k-v in this updated version.  Currently the config is set per-scheduler and the shebang is pulled from the scheduler's k-v config. This is solved if we use the deprecation method I suggested.  Although I think it's best practice for the documentation to show these as set so people understand.

guillaumeeb replied:
I think they shouldn't if they keep the default value, see my response to the other comment. I will try this tomorrow. Default shebang value should be inherited from default config if not specified.

guillaumeeb replied:
And so based on the above discussion, I think we really should remove the majority of examples containing `shebang` in the doc, only keeping a few specific. Maybe add the config you are using in your cluster in the example section?

danpf replied:
I deleted a bunch, I only left it in places where the docs describe the extra args as well!!

guillaumeeb commented on docs/source/examples.rst:
```diff
@@ -23,6 +24,7 @@ PBS Deployments
 
    cluster = PBSCluster(processes=18,
                         threads=4,
+                        shebang='#!/usr/bin/env bash',
```
Still the same, shebang is optional, maybe put it once, but not everywhere.

guillaumeeb commented:
You've got flake8 checks failing here.

willirath commented:
Looks like this is all fine now?

guillaumeeb commented:
Yes ! Thanks to both of you here, merging.
